### ISSUE TITLE
render_loss oscillate

### ISSUE THREAD

**Issue by Haoyanlong** [sklcg/shlight#9]
@sklcg ,hello, I want to use the pytorch  implementing your method.But the render_loss always oscillating.As follows:
my code:
![image](https://user-images.githubusercontent.com/23288056/55458516-c45e4c80-561f-11e9-94a7-83a2dac12596.png)
the visualization:

![image](https://user-images.githubusercontent.com/23288056/55458486-b5779a00-561f-11e9-8287-d0923d5b6c2c.png)
And I set the lr in feature extractor as 1e-5, the lr in dense layer and convs2x3 as 1e-3.Could you help me?

**Comment by linzi-h** [comment]
我个人理解是：
1.作者源码里面的render_loss大多情况下是以envmap来计算的，它相当于用球谐函数重构公式来重构光照分布。如果一直用envmap来算loss，整体趋势应该是稳定的下降。
2.源码里还是会用上特定三维模型的shmap的，那应该是用球谐函数特征来渲染虚拟物体的光照效果，所以整体loss会有一定震荡。
ps:选择shmap的代码（pytorch）
![image](https://user-images.githubusercontent.com/18361553/55677881-b6ac0e00-5922-11e9-8761-d88803f13be3.png)

**Comment by linzi-h** [comment]
我在pytorch里面的实现，用原文的学习率5e-4。
另外我render_loss是这样的：
torch.mean(torch.sum((render_pred -render_gt) ** 2, (1, 2)) / torch.nonzero(torch.sum(shmap, -1)).size(0))

**Comment by Haoyanlong** [comment]
@linzi-h ,thank you very much! When I train the model, the rendered image is randomly selected and the render loss may oscillate.But after I modified it, I still can't realize the rendering effect in the paper. Could you give me some your render result?

**Comment by linzi-h** [comment]
@Haoyanlong ，我修改过网络结构，这是网络预测的渲染效果图
![prediction](https://user-images.githubusercontent.com/18361553/55768966-8b0b5e00-5ab1-11e9-95e7-811d4d87804b.png)

**Comment by linzi-h** [comment]
图片路径：'0019/430_x.jpg'；'0074/409_x.jpg',

**Comment by sklcg** [comment]
@Haoyanlong , Thanks for @linzi-h ‘s reply. As @linzi-h  mentioned，the 3rd shmap is the shmap mainly used for rendering image during traning. It can make loss value converge better. Futhermore,  larger batchsize, SGD optimizer, deeper network model, smaller learning rate or other training tricks can also weaken the oscillation of loss function in some degree. You can apply these tricks reasonably based on your experience and observation.
